pkg/offline/offline.go

```diff
@@ -157,6 +157,146 @@ func WithQueue(filepath string, syncLimit int) (heartbeat.HandleOption, error) {
 	}, nil
 }
 
+// Sync returns a function to send queued heartbeats to the WakaTime API.
+func Sync(filepath string, syncLimit int) func(next heartbeat.Handle) error {
+	return func(next heartbeat.Handle) error {
+		log.Debugf("execute offline sync with file %s", filepath)
+
+		hh, err := popHeartbeats(filepath, syncLimit)
+		if err != nil {
+			return fmt.Errorf("failed to fetch heartbeat from offline queue: %s", err)
+		}
+
+		if len(hh) == 0 {
+			log.Debugln("abort execution, as there are no queued heartbeats ready for sending")
+
+			return nil
+		}
+
+		results, err := next(hh)
+		if err != nil {
+			requeueErr := pushHeartbeats(filepath, hh)
+			if requeueErr != nil {
+				log.Errorf("failed to push heatbeats after api error to queue: %s", requeueErr)
+			}
+
+			return err
+		}
+
+		return handleResults(filepath, results, hh)
+	}
+}
+
+func handleResults(filepath string, results []heartbeat.Result, hh []heartbeat.Heartbeat) error {
+	var (
+		err               error
+		withInvalidStatus []heartbeat.Heartbeat
+	)
+
+	// push heartbeats with invalid result status codes to queue
+	for n, result := range results {
+		if n >= len(hh) {
+			log.Warnln("results from api not matching heartbeats sent")
+			break
+		}
+
+		if result.Status != http.StatusCreated &&
+			result.Status != http.StatusAccepted &&
+			result.Status != http.StatusBadRequest {
+			withInvalidStatus = append(withInvalidStatus, hh[n])
+		}
+	}
+
+	if len(withInvalidStatus) > 0 {
+		log.Debugf("pushing %d heartbeat(s) with invalid result to queue", len(withInvalidStatus))
+
+		err = pushHeartbeats(filepath, withInvalidStatus)
```

Comment:
Maybe the retry goes here instead of in `pushHeartbeats`, up to you.

Reply:
`syncLimit` should just be `limit` because this code has nothing to do with syncing.. it's just popping multiple (up to 24) heartbeats from the offline db.

Reply:
Added a retry. Check new commits :+1: